File: apps/macos/Packages/DailyDoListModels/Sources/DailyDoListModels/Remote.swift
import Foundation

// Swift mirror of the placement, device settings, pairing and always-on machine shapes of
// `packages/core/src/protocol.ts`. Enums a newer daemon may extend are `WireEnum`s.

// MARK: - Placement, readiness

/// Where this device's agent runs (a device-local setting): here and taking the agent over from the
/// always-on machine (`thisDevice`), relayed to the always-on machine (`alwaysOnMachine`), or this
/// is the always-on machine (`alwaysOnHost`). Without sync a daemon runs its own agent anyway.
public struct AgentPlacement: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let thisDevice: Self = "this_device"
  public static let alwaysOnMachine: Self = "always_on_machine"
  public static let alwaysOnHost: Self = "always_on_host"
}

/// The device that holds the agent lease.
public struct AgentRunsOn: Codable, Hashable, Sendable {
  public var deviceId: String
  public var name: String
  public var thisDevice: Bool
  /// The holder requested the lease with priority "host".
  public var alwaysOnMachine: Bool

  public init(deviceId: String, name: String, thisDevice: Bool, alwaysOnMachine: Bool) {
    self.deviceId = deviceId
    self.name = name
    self.thisDevice = thisDevice
    self.alwaysOnMachine = alwaysOnMachine
  }
}

/// The link to the always-on machine's agent.
public struct RelayState: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let off: Self = "off"
  public static let connecting: Self = "connecting"
  public static let connected: Self = "connected"
  public static let unreachable: Self = "unreachable"
  public static let notPaired: Self = "not_paired"
}

/// Why the agent is held on this device despite the stored placement: no always-on machine is set
/// up, or this device doesn't sync.
public struct HeldHereReason: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let noMachine: Self = "no_machine"
  public static let noSync: Self = "no_sync"
}

/// This device's placement and who runs the agent now.
public struct AgentPlacementStatus: Codable, Hashable, Sendable {
  /// The stored choice (see `heldHere` for when it can't apply).
  public var placement: AgentPlacement
  /// Why the agent runs on this device anyway; nil when the stored choice applies.
  public var heldHere: HeldHereReason?
  /// Who runs the agent now (nil: nobody, or unknown without sync).
  public var runsOn: AgentRunsOn?
  public var relay: RelayState
  /// Short, human ("Taking over from vm-1…", "Handing the agent to vm-1…").
  public var note: String?

  public init(
    placement: AgentPlacement, heldHere: HeldHereReason? = nil, runsOn: AgentRunsOn?,
    relay: RelayState, note: String? = nil
  ) {
    self.placement = placement
    self.heldHere = heldHere
    self.runsOn = runsOn
    self.relay = relay
    self.note = note
  }

  enum CodingKeys: String, CodingKey { case placement, heldHere, runsOn, relay, note }

  public func encode(to encoder: Encoder) throws {
    var c = encoder.container(keyedBy: CodingKeys.self)
    try c.encode(placement, forKey: .placement)
    try c.encodeIfPresent(heldHere, forKey: .heldHere)
    // Required on the wire: null when nobody runs the agent.
    try c.encode(runsOn, forKey: .runsOn)
    try c.encode(relay, forKey: .relay)
    try c.encodeIfPresent(note, forKey: .note)
  }
}

/// Whether desktop control can run on a daemon's machine.
public struct ComputerReadiness: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let available: Self = "available"
  public static let needsPermissions: Self = "needs_permissions"
  public static let unsupported: Self = "unsupported"
}

/// Whether a daemon can run the agent: its harness, a model credential (never the value), the
/// browser, desktop control and connectors.
public struct AgentReadiness: Codable, Hashable, Sendable {
  public struct Harness: Codable, Hashable, Sendable {
    /// The harness's wire value, kept as sent so a harness a newer daemon added still decodes.
    public var kind: String
    public var ready: Bool
    public var problem: String?

    public init(kind: String, ready: Bool, problem: String? = nil) {
      self.kind = kind
      self.ready = ready
      self.problem = problem
    }

    public init(kind: AgentHarnessKind, ready: Bool, problem: String? = nil) {
      self.init(kind: kind.rawValue, ready: ready, problem: problem)
    }

    /// The harness, when this client knows it.
    public var knownKind: AgentHarnessKind? { AgentHarnessKind(rawValue: kind) }
  }

  public struct Connectors: Codable, Hashable, Sendable {
    public var configured: Int
    public var connected: Int

    public init(configured: Int, connected: Int) {
      self.configured = configured
      self.connected = connected
    }
  }

  public var harness: Harness
  /// A model credential for the configured harness is present.
  public var modelCredential: Bool
  public var browser: Bool
  public var computer: ComputerReadiness
  public var connectors: Connectors

  public init(
    harness: Harness, modelCredential: Bool, browser: Bool, computer: ComputerReadiness,
    connectors: Connectors
  ) {
    self.harness = harness
    self.modelCredential = modelCredential
    self.browser = browser
    self.computer = computer
    self.connectors = connectors
  }
}

// MARK: - This daemon's device-local settings

/// This device's link to the sync service. The vault token is never returned.
public struct DeviceSyncSetup: Codable, Hashable, Sendable {
  /// nil: not syncing with the sync service.
  public var url: String?
  public var vault: String?
  /// A vault token is saved.
  public var hasToken: Bool

  public init(url: String?, vault: String?, hasToken: Bool) {
    self.url = url
    self.vault = vault
    self.hasToken = hasToken
  }

  enum CodingKeys: String, CodingKey { case url, vault, hasToken }

  public func encode(to encoder: Encoder) throws {
    var c = encoder.container(keyedBy: CodingKeys.self)
    // Required on the wire: null when not syncing.
    try c.encode(url, forKey: .url)
    try c.encode(vault, forKey: .vault)
    try c.encode(hasToken, forKey: .hasToken)
  }
}

/// A device setting an environment variable can fix (`DeviceSettingsResponse.lockedByEnv`).
public struct DeviceSettingField: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let placement: Self = "placement"
  public static let remoteHosts: Self = "remoteHosts"
  public static let sync: Self = "sync"
}

/// `GET /api/device`: this daemon's device-local settings (never synced).
public struct DeviceSettingsResponse: Codable, Hashable, Sendable {
  public struct Device: Codable, Hashable, Sendable {
    public var id: String
    public var name: String

    public init(id: String, name: String) {
      self.id = id
      self.name = name
    }
  }

  public var device: Device
  public var placement: AgentPlacement
  /// Names this daemon answers to besides loopback (e.g. its tailnet name), lowercase.
  public var remoteHosts: [String]
  public var sync: DeviceSyncSetup
  /// Fields set by environment variables; shown read-only.
  public var lockedByEnv: [DeviceSettingField]

  public init(
    device: Device, placement: AgentPlacement, remoteHosts: [String], sync: DeviceSyncSetup,
    lockedByEnv: [DeviceSettingField] = []
  ) {
    self.device = device
    self.placement = placement
    self.remoteHosts = remoteHosts
    self.sync = sync
    self.lockedByEnv = lockedByEnv
  }

  public func isLocked(_ field: DeviceSettingField) -> Bool { lockedByEnv.contains(field) }
}

/// Body of `PATCH /api/device`: only non-nil fields are sent.
public struct DeviceSettingsPatch: Codable, Hashable, Sendable {
  /// 1–64 characters, trimmed.
  public var name: String?
  public var placement: AgentPlacement?
  /// DNS names with an optional `:port`, at most 8 (no IPs, schemes or paths).
  public var remoteHosts: [String]?

  public init(name: String? = nil, placement: AgentPlacement? = nil, remoteHosts: [String]? = nil) {
    self.name = name
    self.placement = placement
    self.remoteHosts = remoteHosts
  }
}

/// Body of `PUT /api/device/sync`.
public struct DeviceSyncSetupRequest: Codable, Hashable, Sendable {
  /// https (plain http only to loopback).
  public var url: String
  public var vault: String
  /// nil keeps the saved token.
  public var token: String?

  public init(url: String, vault: String, token: String? = nil) {
    self.url = url
    self.vault = vault
    self.token = token
  }
}

// MARK: - Pairing

/// What a paired device is: a browser (HttpOnly cookie), a native app, or another daemon.
public struct PairedDeviceKind: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let browser: Self = "browser"
  public static let app: Self = "app"
  public static let daemon: Self = "daemon"
}

/// A device holding a credential for this daemon.
public struct PairedDevice: Codable, Hashable, Sendable, Identifiable {
  public var id: String
  public var name: String
  public var kind: PairedDeviceKind
  public var createdAt: EpochMillis
  public var lastSeenAt: EpochMillis?
  /// The device making this request.
  public var current: Bool?

  public init(
    id: String, name: String, kind: PairedDeviceKind, createdAt: EpochMillis,
    lastSeenAt: EpochMillis?, current: Bool? = nil
  ) {
    self.id = id
    self.name = name
    self.kind = kind
    self.createdAt = createdAt
    self.lastSeenAt = lastSeenAt
    self.current = current
  }

  enum CodingKeys: String, CodingKey { case id, name, kind, createdAt, lastSeenAt, current }

  public func encode(to encoder: Encoder) throws {
    var c = encoder.container(keyedBy: CodingKeys.self)
    try c.encode(id, forKey: .id)
    try c.encode(name, forKey: .name)
    try c.encode(kind, forKey: .kind)
    try c.encode(createdAt, forKey: .createdAt)
    // Required on the wire: null before first use.
    try c.encode(lastSeenAt, forKey: .lastSeenAt)
    try c.encodeIfPresent(current, forKey: .current)
  }
}

/// Body of `POST /api/pairing-codes`.
public struct PairingCodeRequest: Codable, Hashable, Sendable {
  public var name: String?
  public init(name: String? = nil) { self.name = name }
}

/// A single-use pairing code.
public struct PairingCodeResponse: Codable, Hashable, Sendable {
  /// 8 characters without look-alikes; show it as `displayCode`.
  public var code: String
  public var expiresAt: EpochMillis
  /// `https://<first remote host>` for a QR code; nil without remote hosts.
  public var url: String?

  public init(code: String, expiresAt: EpochMillis, url: String?) {
    self.code = code
    self.expiresAt = expiresAt
    self.url = url
  }

  /// `XXXX-XXXX`.
  public var displayCode: String {
    guard code.count == 8 else { return code }
    return "\(code.prefix(4))-\(code.suffix(4))"
  }

  enum CodingKeys: String, CodingKey { case code, expiresAt, url }

  public func encode(to encoder: Encoder) throws {
    var c = encoder.container(keyedBy: CodingKeys.self)
    try c.encode(code, forKey: .code)
    try c.encode(expiresAt, forKey: .expiresAt)
    // Required on the wire: null without remote hosts.
    try c.encode(url, forKey: .url)
  }
}

/// Body of `POST /api/pair`: exchanges a pairing code (typed as shown, any case) for a credential.
public struct PairRequest: Codable, Hashable, Sendable {
  public var code: String
  public var name: String
  public var kind: PairedDeviceKind

  public init(code: String, name: String, kind: PairedDeviceKind) {
    self.code = code
    self.name = name
    self.kind = kind
  }
}

/// The paired device and, for `app` and `daemon` kinds, its token (shown once).
public struct PairResponse: Codable, Hashable, Sendable {
  public var device: PairedDevice
  public var token: String?

  public init(device: PairedDevice, token: String? = nil) {
    self.device = device
    self.token = token
  }
}

/// `GET /api/devices`.
public struct PairedDevicesResponse: Codable, Hashable, Sendable {
  public var devices: [PairedDevice]
  public init(devices: [PairedDevice]) { self.devices = devices }
}

// MARK: - The always-on machine

/// The always-on machine from this device's side: its address, this device's pairing and what the
/// machine reports.
public struct MachineStatusResponse: Codable, Hashable, Sendable {
  /// Where the agent runs, as the machine reports it.
  public struct Agent: Codable, Hashable, Sendable {
    public var runsOn: AgentRunsOn?
    public var problem: String?

    public init(runsOn: AgentRunsOn?, problem: String? = nil) {
      self.runsOn = runsOn
      self.problem = problem
    }

    enum CodingKeys: String, CodingKey { case runsOn, problem }

    public func encode(to encoder: Encoder) throws {
      var c = encoder.container(keyedBy: CodingKeys.self)
      try c.encode(runsOn, forKey: .runsOn)
      try c.encodeIfPresent(problem, forKey: .problem)
    }
  }

  /// nil: no always-on machine configured.
  public var machine: AlwaysOnMachine?
  /// This device holds a credential for the machine.
  public var paired: Bool
  /// nil: not checked yet, or no machine.
  public var reachable: Bool?
  public var checkedAt: EpochMillis?
  public var version: String?
  public var agent: Agent?
  public var readiness: AgentReadiness?
  public var error: String?

  public init(
    machine: AlwaysOnMachine?, paired: Bool, reachable: Bool?, checkedAt: EpochMillis?,
    version: String? = nil, agent: Agent? = nil, readiness: AgentReadiness? = nil,
    error: String? = nil
  ) {
    self.machine = machine
    self.paired = paired
    self.reachable = reachable
    self.checkedAt = checkedAt
    self.version = version
    self.agent = agent
    self.readiness = readiness
    self.error = error
  }

  enum CodingKeys: String, CodingKey {
    case machine, paired, reachable, checkedAt, version, agent, readiness, error
  }

  public func encode(to encoder: Encoder) throws {
    var c = encoder.container(keyedBy: CodingKeys.self)
    // Required on the wire: null without a machine or before a check.
    try c.encode(machine, forKey: .machine)
    try c.encode(paired, forKey: .paired)
    try c.encode(reachable, forKey: .reachable)
    try c.encode(checkedAt, forKey: .checkedAt)
    try c.encodeIfPresent(version, forKey: .version)
    try c.encodeIfPresent(agent, forKey: .agent)
    try c.encodeIfPresent(readiness, forKey: .readiness)
    try c.encodeIfPresent(error, forKey: .error)
  }
}

/// Body of `POST /api/machine/pair`.
public struct MachinePairRequest: Codable, Hashable, Sendable {
  /// `https://<tailnet name>[:port]`, no path, query or credentials.
  public var url: String
  public var code: String
  /// nil: the first label of the host.
  public var name: String?

  public init(url: String, code: String, name: String? = nil) {
    self.url = url
    self.code = code
    self.name = name
  }
}

File: packages/contract/src/wire/remote.ts
/**
 * Where the agent runs (placement, readiness), this daemon's device-local settings, pairing other
 * devices with it, and the always-on machine seen from this device.
 */
import {
  isMachineUrl,
  isPairingCode,
  isRemoteHost,
  isSecureServiceUrl,
  normalizeMachineUrl,
  normalizePairingCode,
  normalizeRemoteHost,
  PAIRING_CODE_ALPHABET,
  PAIRING_CODE_LENGTH,
  REMOTE_LIMITS,
  SYNC_ID_PATTERN,
} from "@ddl/core";
import { z } from "zod";
import {
  CountSchema,
  DeviceNameInputSchema,
  DeviceNameSchema,
  EpochMsSchema,
  RuntimeIdSchema,
  SyncDeviceNameSchema,
} from "./primitives";
import { named } from "./registry";
import { AgentHarnessKindSchema, AlwaysOnMachineSchema } from "./settings";

/** Device ids of the sync service (`dev_…`): 1–64 URL-safe characters. */
export const SyncDeviceIdSchema = z
  .string()
  .regex(SYNC_ID_PATTERN, "must be 1-64 characters of A-Z a-z 0-9 _ -")
  .describe("Device id (as the sync service knows it).");

const VersionSchema = z.string().min(1).max(100);

// ── Placement, readiness ─────────────────────────────────────────────────

export const AgentPlacementSchema = named(
  "AgentPlacement",
  "Where this device's agent runs: `this_device` (here; takes the agent over from the always-on machine), `always_on_machine` (never here; relayed to the always-on machine) or `always_on_host` (this is the always-on machine; it runs the agent when no `this_device` does). Without sync a daemon runs its own agent whatever the placement.",
  z.enum(["this_device", "always_on_machine", "always_on_host"]),
);

export const AgentRunsOnSchema = named(
  "AgentRunsOn",
  "The device that holds the agent lease.",
  z.looseObject({
    deviceId: SyncDeviceIdSchema,
    name: SyncDeviceNameSchema,
    thisDevice: z.boolean(),
    alwaysOnMachine: z.boolean().describe('The holder requested the lease with priority "host".'),
  }),
);

export const RelayStateSchema = named(
  "RelayState",
  "The link to the always-on machine's agent: `off` (not relaying), `connecting`, `connected`, `unreachable` or `not_paired`.",
  z.enum(["off", "connecting", "connected", "unreachable", "not_paired"]),
);

export const AgentPlacementStatusSchema = named(
  "AgentPlacementStatus",
  "This device's placement and who runs the agent now.",
  z.looseObject({
    placement: AgentPlacementSchema.describe(
      "The stored choice (see `heldHere` for when it can't apply).",
    ),
    heldHere: z
      .enum(["no_machine", "no_sync"])
      .optional()
      .describe(
        "Why the agent is held on this device despite the stored choice: no always-on machine is set up (`no_machine`), or this device doesn't sync (`no_sync`).",
      ),
    runsOn: AgentRunsOnSchema.nullable().describe(
      "Who runs the agent now (null: nobody, or unknown without sync).",
    ),
    relay: RelayStateSchema,
    note: z
      .string()
      .optional()
      .describe('Short, human ("Taking over from vm-1…", "Handing the agent to vm-1…").'),
  }),
);

export const AgentReadinessSchema = named(
  "AgentReadiness",
  "Whether a daemon can run the agent: its harness, a model credential (never the value), the browser, desktop control and connectors.",
  z.looseObject({
    harness: z.looseObject({
      kind: AgentHarnessKindSchema,
      ready: z.boolean(),
      problem: z.string().optional(),
    }),
    modelCredential: z
      .boolean()
      .describe("A model credential for the configured harness is present."),
    browser: z.boolean(),
    computer: z.enum(["available", "needs_permissions", "unsupported"]),
    connectors: z.looseObject({ configured: CountSchema, connected: CountSchema }),
  }),
);

// ── This daemon's device-local settings ──────────────────────────────────

/** A remote host as the daemon reports it: `host[:port]`, lowercase (`normalizeRemoteHost`). */
export const RemoteHostSchema = z
  .string()
  .min(1)
  .max(REMOTE_LIMITS.hostnameLength + 6)
  .refine(isRemoteHost, "must be a lowercase DNS name with an optional :port")
  .describe("`host[:port]`, lowercase DNS name (never an IP address or a loopback name).");

/** A remote host as a client sends it: trimmed and lowercased, then validated. */
export const RemoteHostInputSchema = z
  .string()
  .max(REMOTE_LIMITS.hostnameLength + 16)
  .trim()
  .toLowerCase()
  .refine(
    (host) => normalizeRemoteHost(host) === host,
    "must be a DNS name with an optional :port (no scheme, path, IP address or loopback name)",
  );

const DeviceSettingFieldSchema = z.enum(["placement", "remoteHosts", "sync"]);

export const DeviceSyncSetupSchema = named(
  "DeviceSyncSetup",
  "This device's link to the sync service. The vault token is never returned.",
  z.looseObject({
    url: z
      .string()
      .min(1)
      .max(2048)
      .nullable()
      .describe("null: not syncing with the sync service."),
    vault: SyncDeviceIdSchema.nullable().describe("The sync vault id."),
    hasToken: z.boolean().describe("A vault token is saved."),
  }),
);

export const DeviceSettingsResponseSchema = named(
  "DeviceSettingsResponse",
  "This daemon's device-local settings (kept in `$DDL_HOME`, never synced).",
  z.looseObject({
    device: z.looseObject({ id: SyncDeviceIdSchema, name: SyncDeviceNameSchema }),
    placement: AgentPlacementSchema,
    remoteHosts: z
      .array(RemoteHostSchema)
      .max(REMOTE_LIMITS.remoteHosts)
      .describe("Names this daemon answers to besides loopback (e.g. its tailnet name)."),
    sync: DeviceSyncSetupSchema,
    lockedByEnv: z
      .array(DeviceSettingFieldSchema)
      .describe("Fields set by environment variables; clients show them read-only."),
  }),
);

export const DeviceSettingsPatchSchema = named(
  "DeviceSettingsPatch",
  "Body of `PATCH /api/device`: only the fields to change.",
  z.strictObject({
    name: DeviceNameInputSchema.optional().describe("Trimmed, then 1–64 characters."),
    placement: AgentPlacementSchema.optional(),
    remoteHosts: z
      .array(RemoteHostInputSchema)
      .max(REMOTE_LIMITS.remoteHosts)
      .refine((hosts) => new Set(hosts).size === hosts.length, "must not repeat a host")
      .optional()
      .describe(
        "DNS names with an optional `:port`, at most 8; trimmed and lowercased. No IPs, schemes or paths.",
      ),
  }),
);

export const DeviceSyncSetupRequestSchema = named(
  "DeviceSyncSetupRequest",
  "Body of `PUT /api/device/sync`: point this device at the sync service.",
  z.strictObject({
    url: z
      .string()
      .min(1)
      .max(2048)
      .refine(isSecureServiceUrl, "must use https (plain http only to loopback), no credentials")
      .describe("The sync service: https (plain http only to loopback)."),
    vault: SyncDeviceIdSchema.describe("The vault id printed by `ddl-sync vault create`."),
    token: z
      .string()
      .trim()
      .min(1)
      .max(1024)
      .optional()
      .describe("The vault token; omit to keep the saved one. Stored 0600 in `$DDL_HOME`."),
  }),
);

// ── Pairing ──────────────────────────────────────────────────────────────

/** A pairing code as a user types it: `XXXX-XXXX`, any case, spaces and dashes ignored. */
export const PairingCodeInputSchema = z
  .string()
  .min(1)
  .max(32)
  .refine(
    (code) => normalizePairingCode(code) !== null,
    `must be the ${PAIRING_CODE_LENGTH}-character pairing code (XXXX-XXXX)`,
  )
  .describe(
    `The pairing code: ${PAIRING_CODE_LENGTH} characters of \`${PAIRING_CODE_ALPHABET}\`, any case; spaces and dashes are ignored.`,
  );

export const PairedDeviceKindSchema = named(
  "PairedDeviceKind",
  "`browser` (gets an HttpOnly cookie), `app` (a native client) or `daemon` (another daemon relaying to this one).",
  z.enum(["browser", "app", "daemon"]),
);

export const PairedDeviceSchema = named(
  "PairedDevice",
  "A device holding a credential for this daemon.",
  z.looseObject({
    id: RuntimeIdSchema,
    name: DeviceNameSchema,
    kind: PairedDeviceKindSchema,
    createdAt: EpochMsSchema,
    lastSeenAt: EpochMsSchema.nullable().describe("Last use (updated at most once a minute)."),
    current: z.boolean().optional().describe("The device making this request."),
  }),
);

export const PairingCodeRequestSchema = named(
  "PairingCodeRequest",
  "Body of `POST /api/pairing-codes`.",
  z.strictObject({
    name: DeviceNameInputSchema.optional().describe("What the new device will be called."),
  }),
);

export const PairingCodeResponseSchema = named(
  "PairingCodeResponse",
  "A single-use pairing code.",
  z.looseObject({
    code: z
      .string()
      .length(PAIRING_CODE_LENGTH)
      .refine(isPairingCode, `must use only ${PAIRING_CODE_ALPHABET}`)
      .describe(
        `${PAIRING_CODE_LENGTH} characters of \`${PAIRING_CODE_ALPHABET}\`; show it as XXXX-XXXX.`,
      ),
    expiresAt: EpochMsSchema,
    url: z
      .string()
      .max(300)
      .refine(
        (url) => url.startsWith("https://") && isMachineUrl(url),
        "must be https://<remote host>",
      )
      .nullable()
      .describe("`https://<first remote host>` for a QR code; null without remote hosts."),
  }),
);

export const PairRequestSchema = named(
  "PairRequest",
  "Body of `POST /api/pair`: exchange a pairing code for a device credential.",
  z.strictObject({
    code: PairingCodeInputSchema,
    name: DeviceNameInputSchema,
    kind: PairedDeviceKindSchema,
  }),
);

export const PairResponseSchema = named(
  "PairResponse",
  "The paired device and, for `app` and `daemon` kinds, its token (shown once). A browser gets an HttpOnly cookie instead.",
  z.looseObject({
    device: PairedDeviceSchema,
    token: z
      .string()
      .min(16)
      .max(512)
      .optional()
      .describe("Send as `Authorization: Bearer <token>`."),
  }),
);

export const PairedDevicesResponseSchema = named(
  "PairedDevicesResponse",
  "Every device paired with this daemon.",
  z.looseObject({ devices: z.array(PairedDeviceSchema) }),
);

// ── The always-on machine ────────────────────────────────────────────────

export const MachineStatusResponseSchema = named(
  "MachineStatusResponse",
  "The always-on machine from this device's side: its address, this device's pairing and what the machine reports.",
  z.looseObject({
    machine: AlwaysOnMachineSchema.nullable().describe("null: no always-on machine configured."),
    paired: z.boolean().describe("This device holds a credential for the machine."),
    reachable: z.boolean().nullable().describe("null: not checked yet, or no machine."),
    checkedAt: EpochMsSchema.nullable(),
    version: VersionSchema.optional().describe("The machine's daemon version."),
    agent: z
      .looseObject({
        runsOn: AgentRunsOnSchema.nullable(),
        problem: z.string().optional(),
      })
      .optional()
      .describe("Where the agent runs, as the machine reports it."),
    readiness: AgentReadinessSchema.optional().describe("The machine's readiness."),
    error: z.string().optional().describe("Why the last check or pairing failed."),
  }),
);

export const MachinePairRequestSchema = named(
  "MachinePairRequest",
  "Body of `POST /api/machine/pair`: pair this device with the always-on machine.",
  z.strictObject({
    url: z
      .string()
      .min(1)
      .max(300)
      .refine(
        (url) => normalizeMachineUrl(url) !== null,
        "must be https://<host>[:port] without path, query or credentials (plain http only to loopback)",
      )
      .describe("`https://<tailnet name>[:port]`; case and a trailing `/` are normalized."),
    code: PairingCodeInputSchema,
    name: DeviceNameInputSchema.optional().describe("Default: the first label of the host."),
  }),
);
